content/blog/2020-03-dplyr-1-0-0-and-vctrs/index.Rmarkdown

```diff
@@ -0,0 +1,149 @@
+---
+title: dplyr 1.0.0 and vctrs
+author: Hadley Wickham
+date: '2020-03-25'
+slug: dplyr-1-0-0-and-vctrs
+photo:
+  url: https://unsplash.com/photos/IstXvxHGoA4
+  author: 35mm
+categories:
+  - package
+tags:
+  - dplyr
+---
+
+[dplyr 1.0.0 is coming soon](https://www.tidyverse.org/blog/2020/03/dplyr-1-0-0-is-coming-soon/), and we've already shown you how [`summarise()` is growing](https://www.tidyverse.org/blog/2020/03/dplyr-1-0-0-summarise/). Today, I wanted to talk a little bit how dplyr 1.0.0 now uses the  [vctrs](http://vctrs.r-lib.org/) package. This post explains why vctrs is so important, why we can't just copy what base R does, how to interpret some of new error messages that you'll see, and some of the major changes since the last version.
+
+```{r, include = FALSE}
+knitr::opts_chunk$set(collapse = TRUE, comment = "#>")
+```
+
+## Combining vectors
+
+The heart of the reason we're using vctrs is the need to combine vectors together. You're already familiar with one base R tool for combining vectors, `c()`:
+
+```{r}
+c(1, 2, 3)
+c("a", "b", "c")
+```
+
+But combining vectors comes up in many places in the tidyverse, e.g.:
+
+* `dplyr::mutate()` and `dplyr::summarise()` have to combine the results 
+  from each group.
+* `dplyr::bind_rows()` has to combine columns from different data frames.
+* `dplyr::full_join()`  has to combine the keys from the `x` and `y` data
+  frames.
+* `tidyr::pivot_longer()` has to combine multiple columns into one.
+
+Our goal is to unify the code that underlies all these various functions so that there's one consistent, principled approach. We've already made the change in [tidyr](https://www.tidyverse.org/blog/2019/09/tidyr-1-0-0/), and now it's dplyr's turn. 
+
+## Base R behaviour
+
+You might wonder why we can't just copy the behaviour of `c()`. Unfortunately `c()` has some major downsides:
+
+*   It converts factors to their underlying integer levels:
```

Comment:
First, it isn't implemented for a range of important classes like factors. Providing implementations is now difficult because it would break a lot of historical code.